engine/consensus/approvals/request_tracker.go

```diff
@@ -58,40 +61,66 @@ RequestTracker
 // Index on result ID, incorporated block ID and chunk index.
 // Is concurrency-safe.
 type RequestTracker struct {
+	headers           storage.Headers
 	index             map[flow.Identifier]map[flow.Identifier]map[uint64]RequestTrackerItem
 	blackoutPeriodMin int
 	blackoutPeriodMax int
-	lock              sync.RWMutex
+	lock              sync.Mutex
+	byHeight          map[uint64]flow.IdentifierList
+	lowestHeight      uint64
 }
 
 // NewRequestTracker instantiates a new RequestTracker with blackout periods
 // between min and max seconds.
-func NewRequestTracker(blackoutPeriodMin, blackoutPeriodMax int) *RequestTracker {
+func NewRequestTracker(headers storage.Headers, blackoutPeriodMin, blackoutPeriodMax int) *RequestTracker {
 	return &RequestTracker{
+		headers:           headers,
 		index:             make(map[flow.Identifier]map[flow.Identifier]map[uint64]RequestTrackerItem),
+		byHeight:          make(map[uint64]flow.IdentifierList),
 		blackoutPeriodMin: blackoutPeriodMin,
 		blackoutPeriodMax: blackoutPeriodMax,
 	}
 }
 
-// Get returns the tracker item for a specific chunk, and creates a new one if
-// it doesn't exist.
-func (rt *RequestTracker) Get(resultID, incorporatedBlockID flow.Identifier, chunkIndex uint64) RequestTrackerItem {
-	rt.lock.RLock()
+// TryUpdate tries to update tracker item if it's not in blackout period. Returns the tracker item for a specific chunk
+// (creates it if it doesn't exists) and whenever request item was successfully updated or not.
+// Since RequestTracker prunes items by height it can't accept items for height lower than cached lowest height.
+// If height of executed block pointed by execution result is smaller than the lowest height, sentinel mempool.DecreasingPruningHeightError is returned.
+// In case execution result points to unknown executed block exception will be returned.
+func (rt *RequestTracker) TryUpdate(result *flow.ExecutionResult, incorporatedBlockID flow.Identifier, chunkIndex uint64) (RequestTrackerItem, bool, error) {
```

Comment:
Can we add a test case that concurrently calling `TryUpdate` 10 times, and verify 10 requests are tracked.

Reply:
Implemented.